RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc

```diff
@@ -943,3 +997,40 @@ void ReducedEGProducer::relinkGsfElectronCore(reco::GsfElectron& gsfElectron,
     gsfElectron.setCore(coreref);
   }
 }
+
+void ReducedEGProducer::calibratePhoton(reco::Photon& photon, 
+					const reco::PhotonRef& oldPhoRef,
+					const edm::ValueMap<float>& energyMap,
+					const edm::ValueMap<float>& energyErrMap)
+{
+  float newEnergy = energyMap[oldPhoRef];
+  float newEnergyErr = energyErrMap[oldPhoRef];
+  photon.setCorrectedEnergy(reco::Photon::P4type::regression2, newEnergy, newEnergyErr, true);
+ 
+}
+
+void ReducedEGProducer::calibrateElectron(reco::GsfElectron& electron,
+					  const reco::GsfElectronRef& oldEleRef,
+					  const edm::ValueMap<float>& energyMap,
+					  const edm::ValueMap<float>& energyErrMap,
+					  const edm::ValueMap<float>& ecalEnergyMap,
+					  const edm::ValueMap<float>& ecalEnergyErrMap)
+{
+  const float newEnergy = energyMap[oldEleRef];
+  const float newEnergyErr = energyErrMap[oldEleRef];
+  const float newEcalEnergy = ecalEnergyMap[oldEleRef];
+  const float newEcalEnergyErr = ecalEnergyErrMap[oldEleRef];
+
+  const math::XYZTLorentzVector& oldP4 = electron.p4();
+  const float corr = newEnergy / oldP4.E();
+  
+  electron.setCorrectedEcalEnergy(newEcalEnergy);
```

Comment:
I think that the call in L1024 should be ```const math::XYZTLorentzVector oldP4 = electron.p4();``` without a ref.

the call on L1027 ``` electron.setCorrectedEcalEnergy ``` overrides the value of p4 https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/src/GsfElectron.cc#L184-L186. IIUC, from here we have 
- electron.energy = newEcalEnergy
- corr = newEnergy/oldEnergy
- oldP4.energy = newEcalEnergy

So, newP4 below will have [assuming massless] (unitP3*newEcalEnergy*newEnergy/oldEnergy, newEnergy)

Reply:
yes, this is screwed up, thanks for the catch. Oh the joy of E/gamma code.